### PR TITLE
Feature/budget

### PR DESCRIPTION
**Pull request by K-w-e**
I'm sending you the pull request for the implementation of the Budgets feature, just to keep you informed about the direction I took during development. 

Some points of attention already known:

- I removed the "Add category budget" button from the planning page. Now, if you want to modify the budgets, you need to press "Manage." (I did this for both simplicity and because the flow to follow by pressing that button was not clear).
- I added the ability to delete the budget via swipe. In the UI, there was both that option and the option through a "-" button next to the budget. We need to decide which one to keep.
- When you add a new budget or modify an existing one, provide the option to select only the categories that are not already present. This should be easy to fix, but I haven't found a clean way to do it yet.

For any doubts, questions, or suggestions, feel free to reach out!

### CODE REVIEW

File: lib/model/budget.dart [hunk 1]
```diff
@@ -135,4 +164,16 @@ class BudgetMethods extends SossoldiDatabase {
...
+    final db = await database;
+
+    return await db.delete(budgetTable, where: '${BudgetFields.idCategory} = ?', whereArgs: [id]);
+  }
+
+  //PER TEST
+  Future<int> deleteAll() async {
+    final db = await database;
+    return await db.rawDelete("DELETE FROM $budgetTable");
+  }
```
Comment (mikev-cw) [inline lib/model/budget.dart:178]: Remember to remove this on the final submission, if is a method that you just need to simplify your work and it won't go in the final release

File: lib/pages/planning_page/manage_budget_page.dart [hunk 2]
```diff
@@ -0,0 +1,161 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:sossoldi/providers/budgets_provider.dart';
```
Comment (mikev-cw) [inline lib/pages/planning_page/manage_budget_page.dart:3]: don't use `package:sossoldi` for our own packages, just use the absolute path

File: lib/pages/planning_page/planning_page.dart [hunk 3]
```diff
@@ -1,4 +1,6 @@
 import 'package:flutter/material.dart';
+import 'package:sossoldi/model/budget.dart';
```
Comment (mikev-cw) [inline lib/pages/planning_page/planning_page.dart:2]: don't use `package:sossoldi` for our own packages, just use the absolute path

File: lib/pages/planning_page/widget/budget_card.dart [hunk 4]
```diff
@@ -1,13 +1,18 @@
 import 'package:flutter/material.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:sossoldi/pages/planning_page/manage_budget_page.dart';
+import 'package:sossoldi/providers/transactions_provider.dart';
```
Comment (mikev-cw) [inline lib/pages/planning_page/widget/budget_card.dart:4]: don't use `package:sossoldi` for our own packages, just use the absolute path

File: lib/pages/planning_page/widget/budget_category_selector.dart [hunk 5]
```diff
@@ -0,0 +1,112 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:sossoldi/model/budget.dart';
+import 'package:sossoldi/model/category_transaction.dart';
```
Comment (mikev-cw) [inline lib/pages/planning_page/widget/budget_category_selector.dart:5]: don't use `package:sossoldi` for our own packages, just use the absolute path

**Review by mikev-cw — COMMENTED** [review]
Ok, that's very cool!!!
I just made some minor notes, but I've nothing to say about the overall approach.
Also keep an eye to those brand new conflicts.

About this: 

> I added the ability to delete the budget via swipe. In the UI, there was both that option and the option through a "-" button next to the budget. We need to decide which one to keep.

It seems that in the design file @federicopozzato has planned kind of "edit mode", where user can tap the "-" button to delete a budget, but since you can in fact edit all about budgets when you press "Manage", actually having an extra edit mode can be a bit redundant at this stage?

**Comment by federicopozzato** [comment]
@mikev-cw @K-w-e thanks for pointing out those things!
The redundancy of the actions has been done on purpose, to give the user more freedom, even though I understand what you say about too many actions, and you may be right.
For semplicity let's keep it in the way you did, it's fine!

File: lib/pages/planning_page/manage_budget_page.dart [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        const SizedBox(height: 10),
+        Text(
+            "Your monthly budget will be: ${budgets.isEmpty ? 0 : budgets.fold(0, (sum, e) => sum + e.amountLimit.toInt())}€"),
+        const SizedBox(height: 10),
+        const Divider(height: 1, color: grey2),
+        Padding(
+            padding: const EdgeInsets.all(8),
+            child: SizedBox(
+                width: double.infinity,
+                child: ElevatedButton(
```
Comment (theperu) [inline lib/pages/planning_page/manage_budget_page.dart:144]: Could you make the button style similar to the other ones in the app? For example the button to add/edit a transaction.

File: lib/pages/planning_page/manage_budget_page.dart [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                ),
+                Text(
+                  'Amount',
+                  textAlign: TextAlign.right,
+                ),
+              ],
+            )),
+        Expanded(
+            child: Column(children: [
+          ListView.builder(
```
Comment (theperu) [inline lib/pages/planning_page/manage_budget_page.dart:75]: Below the list of categories we should have a text that explains to the user that if they want to delete a budget they should swipe it left or right.

**Review by theperu — CHANGES_REQUESTED** [review]
I really like the implementation, good job! I left a couple of comments for minor changes but apart from that it seems fine to me.
Since you have already done the work on the budgets could you also work on the graphs that are at the bottom of the Dashboard next? 

FYI I just noticed that there are also some conflicts that need to be resolved before we can merge

**Comment by K-w-e** [comment]
PR updated!
That file with conflicts is caused by the fact that that widget has been rearranged and almost completely modified, I don't think I can resolve it myself. You should accept the changes from my branch through the PR.

Let me know if i can help

@theperu

File: lib/pages/planning_page/manage_budget_page.dart (hunk 6 above)
Comment (K-w-e) [reply lib/pages/planning_page/manage_budget_page.dart]: Fixed

File: lib/pages/planning_page/manage_budget_page.dart (hunk 7 above)
Comment (K-w-e) [reply lib/pages/planning_page/manage_budget_page.dart:75]: Added

**Pull request merged**